### ISSUE TITLE
multiobjective_optimization.ipynb SOBOL iterations runs slow ending up with HV=0

### ISSUE THREAD

**Issue by hanochk** [facebook/Ax#732]
Hi Im running MO = 2 objectives, 9 dimensions, however using the MOO_NEHVI instead of the vanilla MOO_EHVI doesn't improve throughput or the fact that the SOBOL runs in ~30 min . Is there a way to accelerate that ? with additional arms for instance? Can you give a pointer to implementation ? the MOO_NEHVI is expected to run in parallel, right ? The fact that SOBOL iterations ending up with HV=0 is normal ? 
Thanks 
Hanoch

**Comment by Balandat** [comment]
That the Sobol approach ends up with zero HV improvement over the reference point here is not too surprising, after all it's just doing random exploration which is generally not very effective. What is surprising is that this takes ~30 minutes to run. For me this takes around 30 seconds on my laptop. Did you change the tutorial in any way (e.g. update the evaluation function)? What environment are you running this in?

**Comment by hanochk** [comment]
Copy-paste the code that maximizes compression with constraints. BTW I've
replaced the  get_MOO_EHVI with the  get_MOO_NEHVI to watch for improvement
but none. Thanks

import pandas as pd
from ax import *

import numpy as np

from ax.metrics.noisy_function import NoisyFunctionMetric
from ax.service.utils.report_utils import exp_to_df
from ax.runners.synthetic import SyntheticRunner

from ax.service.ax_client import AxClient
from ax.service.utils.instantiation import ObjectiveProperties

import torch

# Plotting imports and initialization
from ax.utils.notebook.plotting import render, init_notebook_plotting
from ax.plot.pareto_utils import compute_posterior_pareto_frontier
from ax.plot.pareto_frontier import plot_pareto_frontier
init_notebook_plotting()

# Factory methods for creating multi-objective optimization modesl.
from ax.modelbridge.factory import get_MOO_EHVI, get_MOO_PAREGO

# Analysis utilities, including a method to evaluate hypervolumes
from ax.modelbridge.modelbridge_utils import observed_hypervolume


from botorch.test_functions.base import (
    ConstrainedBaseTestProblem,
    MultiObjectiveTestProblem,
)
from ax.metrics.noisy_function import GenericNoisyFunctionMetric #TODO:
revert to non noisy

from pylab import *
import sys

import M
import torch as nn
import os
from math import sqrt, log

#!g++ -DFSE_VERBOSE -Ofast ../cimpl/main.cc -lm  -shared -o
../HMS/ans_cmprs.so
!g++ -Ofast ../cimpl/main.cc -lm  -shared -o ../HMS/ans_cmprs.so # no
verbose

compress_decompress = M.load_c_compression_module()
M.path_to_error_eval_images = '/mnt/users/images/**/*.jpg'
ax_client = AxClient()

images = M.Images()
minsize = 10 #1048572 #  HK : only one layer FC 1K
model_path='./ml_video_edit_art_generate.onnx'

class max_compress_min_err_moo(MultiObjectiveTestProblem):
    def __init__(self, model_path, images, minsize,
error_method='rel_error',noise_std=None, negate=False):

        self.error_method = error_method
        self.ob_original = M.OnnxBridge(model_path)
        self.ob_q = M.OnnxBridge(model_path)

        self.minsize = minsize

        self.images = images
        self.original_outs = [self.ob_original(i) for i in self.images]

        self.ws = self.ob_original.get_weights()

        self.dim = 0
        for k, w in self.ws.items():
            if len(w.shape) > 1 and w.size > self.minsize:
                self.dim += 1

        self.qws = {}

        self.num_objective = 2
        self._ref_point: float = [0, 0]
#         self._ref_point: float = [6.1876395, -0.048405603] # near best
point by heuristics#[6.238192428550052, 0.044830203]
        self._max_hv = 1

        self._bounds = [(0.0, 0.05)] * self.dim
#         _ref_point: float = [0, 0]
#         _bounds = [(0.0, 10), (0.0, 0.1)] # 1st objective -compression,
2nd is rel_error

        super().__init__(noise_std=noise_std, negate=negate)


    def forward(self, X, noise=True):
#         print("forward is ruuning")

#         raise ValueError("Not imp max_compress_min_err_moo forward")
#         print(X.size == self.dim[0])
#         print(X.size )
#         print(self.dim)

        f_0 = torch.tensor(self._taregt_fun_scale_heuristic_opt(X))
        f_1 = torch.tensor(self._constraint_fun(X))
#         print(f_0, f_1)
        return torch.stack([f_0, f_1], dim=-1)

    def evaluate_true(self, X):
#         print("evaluate_true is ruuning")
        #Evaluate the function (w/o observation noise) on a set of points.
#         print(X.size == self.dim[0])
        f_0 = self._taregt_fun_scale_heuristic_opt(X)
        f_1 = self._constraint_fun(X)
        return torch.stack([f_0, f_1], dim=-1)

    def _taregt_fun_scale_heuristic_opt(self, x):

        original_size = 0
        compressed_size = 0
        ix = 0
        for k, w in (self.ws.items()):
            if len(w.shape) > 1 and w.size > self.minsize:
#                 target_error = (1e-2 + sqrt(w.size) / 2.51e4 )
                target_error = x[ix]
                ix += 1
#                 print("Layer target_error {} size
{}".format(target_error, w.size))
                target_error = np.array(target_error.cpu(),
dtype=np.float32)
                measurements = nn.zeros(5,
dtype=nn.float32).contiguous().numpy()
                weight = nn.tensor(w, dtype=nn.float32).contiguous()
                sz = nn.tensor(weight.shape,
dtype=nn.int32).contiguous().numpy()
                dims = nn.tensor(sz.size,
dtype=nn.int32).contiguous().numpy()
                self.qw = nn.zeros_like(weight,
dtype=nn.float32).contiguous().numpy()
                compress_decompress(weight.numpy(), sz, dims, target_error,
self.qw, measurements)
                self.qws[k] = self.qw.copy()
                compressed_size += measurements[2]
                original_size += weight.numel() * 4.0  # 4bytes per float
#                 print("layer {} quant_err {:0.3f} Compressed Size {:0.1f}
ratio {:0.3f}".format(k, measurements[1], measurements[2], measurements[3]))
            else:
                weight = nn.tensor(w, dtype=nn.float32)
                #         print(weight.shape)
                compressed_size += weight.numel() * 4.0
                original_size += weight.numel() * 4.0  # 4bytes per float
                self.qws[k] = w.copy()

#         print('Total compression ratio {} Scale {}:'.format(original_size
/ compressed_size, x[0]))
        return -(original_size / compressed_size) # Opt : max

    def _constraint_fun(self, x):
#         print("_constraint_fun is running")
#         if self.qws:
#             print(np.unique(self.qws[(24, 'fc.weight')].flatten(),
return_counts=True)[1].shape) # for minsize = 1048576 - 4#([qws[(26,
'vgg.conv1.weight')].min(), qws[(26, 'vgg.conv1.weight')].max()])
#         else: # for optmizers run the constraint() first, then run the
optimizer to get initial func
#             raise ValueError("Not imp _constraint_fun()")
            #taregt_fun_opt(x, div_by_scale_frac=div_by_scale_frac) #
COBYLA runs the constr() first so need to initi the  qws
        # warnings.filterwarnings("ignore")
        self.ob_q.set_weights(self.qws)
        # warnings.filterwarnings("error")

        outs = [self.ob_q(i) for i in self.images]
        errs = []
        for ts, ps in zip(self.original_outs, outs):
            for t, p in zip(ts, ps):
                # errs.append(M.flat_dif(p, t))
                if self.error_method == 'rel_error':
                    errs.append(M.flat_dif(p, t))
                elif self.error_method == 'atan_error':
                    atan_err = M.atan_error(p, t)
                    if atan_err < 0:
                        raise
                    errs.append(atan_err)
                elif self.error_method == 'mse_error':
                    errs.append(M.mse_error(p, t))
                else:
                    raise ValueError("Undefined error method")

        error_metric = np.mean(errs)
#         print('error:', error_metric)
        return error_metric # opt min


tkwargs = {
    "dtype": torch.double,
#     "device": device, # HK
    "device": torch.device( "cpu"), #Force CPU
}

objective_ans = max_compress_min_err_moo(model_path=model_path,
images=images,
                                     minsize=minsize,
negate=True).to(**tkwargs)


search_space = SearchSpace(
    parameters=[RangeParameter(name=f"x{i}", lower=0, upper=0.1,
parameter_type=ParameterType.FLOAT) for i in range(objective_ans.dim)],
)
param_names = [f"x{i}" for i in range(objective_ans.dim)]


def f1(x) -> float:
        x_sorted = [x[p_name] for p_name in param_names]
        return float(objective_ans(torch.tensor(x_sorted,
**tkwargs).clamp(0., 1.))[0]) #HK maybe put clamp to 0.1

def f2(x) -> float:
        x_sorted = [x[p_name] for p_name in param_names]
        return float(objective_ans(torch.tensor(x_sorted,
**tkwargs).clamp(0., 1.))[1])

metric_a = GenericNoisyFunctionMetric("a", f=f1, noise_sd=0.0,
lower_is_better=False)# compression
metric_b = GenericNoisyFunctionMetric("b", f=f2, noise_sd=0.0,
lower_is_better=False)# rel err

mo = MultiObjective(
    objectives=[Objective(metric=metric_a), Objective(metric=metric_b)],
)

objective_thresholds = [
    ObjectiveThreshold(metric=metric, bound=val, relative=False)
    for metric, val in zip(mo.metrics, objective_ans.ref_point)
]

optimization_config = MultiObjectiveOptimizationConfig(
    objective=mo,
    objective_thresholds=objective_thresholds,
)

N_INIT = 6
N_BATCH = 25

def build_experiment():
    experiment = Experiment(
        name="pareto_experiment",
        search_space=search_space,
        optimization_config=optimization_config,
        runner=SyntheticRunner(),
    )
    return experiment

def initialize_experiment(experiment):
    sobol = Models.SOBOL(search_space=experiment.search_space, seed=1234)

    for _ in range(N_INIT):
        experiment.new_trial(sobol.gen(1)).run()

    return experiment.fetch_data()

sobol_experiment = build_experiment()
sobol_data = initialize_experiment(sobol_experiment)

sobol_model = Models.SOBOL(
    experiment=sobol_experiment,
    data=sobol_data,
)
sobol_hv_list = []
for i in range(N_BATCH):
    generator_run = sobol_model.gen(1)
    trial = sobol_experiment.new_trial(generator_run=generator_run)
    trial.run()
    exp_df = exp_to_df(sobol_experiment)
    outcomes = np.array(exp_df[['a', 'b']], dtype=np.double)
    # Fit a GP-based model in order to calculate hypervolume.
    # We will not use this model to generate new points.
    dummy_model = get_MOO_EHVI(
        experiment=sobol_experiment,
        data=sobol_experiment.fetch_data(),
    )
    try:
        hv = observed_hypervolume(modelbridge=dummy_model)
    except:
        hv = 0
        print("Failed to compute hv")
    sobol_hv_list.append(hv)
    print(f"Iteration: {i}, HV: {hv}")

sobol_outcomes = np.array(exp_to_df(sobol_experiment)[['a', 'b']],
dtype=np.double)

ehvi_experiment = build_experiment()
ehvi_data = initialize_experiment(ehvi_experiment)

BATCH_SIZE = 4 # HK todo optimize
ehvi_hv_list = []
ehvi_model = None
for i in range(N_BATCH):
    ehvi_model = get_MOO_EHVI(
        experiment=ehvi_experiment,
        data=ehvi_data,
    )
    generator_run = ehvi_model.gen(1) # why not generator_run =
model.gen(BATCH_SIZE)
    trial = ehvi_experiment.new_trial(generator_run=generator_run)
    trial.run()
    ehvi_data = Data.from_multiple_data([ehvi_data, trial.fetch_data()])

    exp_df = exp_to_df(ehvi_experiment)
    outcomes = np.array(exp_df[['a', 'b']], dtype=np.double)
    try:
        hv = observed_hypervolume(modelbridge=ehvi_model)
    except:
        hv = 0
        print("Failed to compute hv")
    ehvi_hv_list.append(hv)
    print(f"Iteration: {i}, HV: {hv}")

ehvi_outcomes = np.array(exp_to_df(ehvi_experiment)[['a', 'b']],
dtype=np.double)

**Comment by lena-kashtelyan** [comment]
cc @sdaulton on this one

**Comment by lena-kashtelyan** [comment]
Hi @hanochk, could you please post a properly formatted version of your code snippet as a Jupyter notebook so we can look into this for you. It should be:
1. without commented-out code,
2. with outputs shown, so we can see the HV of 0 etc.

If you need them, instructions here for ways of posting a Jupyter notebook with outputs shown: https://github.com/zonca/zonca-blog/blob/master/_posts/2020-09-15-how-to-share-jupyter-notebooks.md).

**Comment by lena-kashtelyan** [comment]
Closing this issue for now as inactive, feel free to reopen when it has the needed updates!
